hazelcast/src/test/java/com/hazelcast/query/impl/predicates/BetweenVisitorTest.java

```diff
@@ -0,0 +1,208 @@
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import static com.hazelcast.query.Predicates.*;
+import static com.hazelcast.query.impl.TypeConverters.INTEGER_CONVERTER;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Matchers.anyString;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.hamcrest.Matchers.*;
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category({QuickTest.class, ParallelTest.class})
+public class BetweenVisitorTest {
+
+    private BetweenVisitor visitor;
+    private Indexes mockIndexes;
+    private Index mockIndex;
+
+    @Before
+    public void setUp() {
+        mockIndexes = mock(Indexes.class);
+        mockIndex = mock(Index.class);
+        when(mockIndexes.getIndex(anyString())).thenReturn(mockIndex);
+        visitor = new BetweenVisitor();
+        useConverter(INTEGER_CONVERTER);
+    }
+
+    @Test
+    public void whenGreaterOrEqualsThanXandLessOrEqualsThanY_thenRewriteToBetweenXandY() {
+        //(age >= 5 and age <= 6)  --> (age between 5 6)
+        Predicate left = greaterEqual("attribute", 5);
+        Predicate right = lessEqual("attribute", 6);
+        Predicate and = and(left, right);
+
+        Predicate result = visitor.visit((AndPredicate) and, mockIndexes);
+        assertBetweenPredicate(result, 5, 6);
+    }
+
+    @Test
+    public void whenPredicatesOutsideTheRangeFromRightExist_thenEliminateThem() {
+        //(age >= 5 and age <= 6 and age < 10)  -->  (age between 5 6)
+        Predicate left = greaterEqual("attribute", 5);
+        Predicate right = lessEqual("attribute", 6);
+        Predicate other = lessEqual("attribute", 10);
+        Predicate and = and(left, right, other);
+
+        Predicate result = visitor.visit((AndPredicate) and, mockIndexes);
+        assertBetweenPredicate(result, 5, 6);
+    }
+
+    @Test
+    public void whenPredicatesOutsideTheRangeFromLeftExist_thenEliminateThem() {
+        //(age >= 5 and age <= 6 and age > 4)  -->  (age between 5 6)
+        Predicate left = greaterEqual("attribute", 5);
+        Predicate right = lessEqual("attribute", 6);
+        Predicate other = greaterEqual("attribute", 4);
+        Predicate and = and(left, right, other);
+
+        Predicate result = visitor.visit((AndPredicate) and, mockIndexes);
+        assertBetweenPredicate(result, 5, 6);
+    }
+
+    @Test
+    public void whenBothBoundariesAreSame_thenRewriteItAsEquals() {
+        //(age >= 5 and age <= 5)  -->  (age = 5)
+        Predicate left = greaterEqual("attribute", 5);
+        Predicate right = lessEqual("attribute", 5);
+        Predicate and = and(left, right);
+
+        EqualPredicate p = (EqualPredicate) visitor.visit((AndPredicate) and, mockIndexes);
+        assertEquals("attribute", p.attribute);
+        assertEquals(5, p.value);
+    }
+
```

Comment:
Would be great to have a test that covers optimization of the `between` operation when the operator is the same in both cases, like
`age > 5 and age > 8`